examples/vision/swin_transformers.py

```diff
@@ -0,0 +1,544 @@
+        self.pos_embed = layers.Embedding(input_dim=num_patch, output_dim=embed_dim)
+
+    def call(self, patch):
+        pos = tf.range(start=0, limit=self.num_patch, delta=1)
+        return self.proj(patch) + self.pos_embed(pos)
+
+
+class PatchMerging(tf.keras.layers.Layer):
+    def __init__(self, num_patch, embed_dim):
+        super(PatchMerging, self).__init__()
+        self.num_patch = num_patch
+        self.embed_dim = embed_dim
+        self.linear_trans = layers.Dense(2 * embed_dim, use_bias=False)
+
+    def call(self, x):
+        H, W = self.num_patch
+        _, _, C = x.get_shape().as_list()
+        x = tf.reshape(x, shape=(-1, H, W, C))
+        x0 = x[:, 0::2, 0::2, :]
+        x1 = x[:, 1::2, 0::2, :]
+        x2 = x[:, 0::2, 1::2, :]
+        x3 = x[:, 1::2, 1::2, :]
+        x = tf.concat((x0, x1, x2, x3), axis=-1)
+        x = tf.reshape(x, shape=(-1, (H // 2) * (W // 2), 4 * C))
+        return self.linear_trans(x)
+
+
+"""
+### Build the model
+
+We will now put together the Swin Transformer model.
+"""
+
+input = layers.Input(input_shape)
+x = PatchExtract(patch_size)(input)
+x = PatchEmbedding(num_patch_x * num_patch_y, embed_dim)(x)
+x = SwinTransformer(
+    dim=embed_dim,
+    num_patch=(num_patch_x, num_patch_y),
+    num_heads=num_heads,
+    window_size=window_size,
+    shift_size=0,
+    num_mlp=num_mlp,
+    qkv_bias=qkv_bias,
+    drop_rate=drop_rate,
+)(x)
+x = SwinTransformer(
+    dim=embed_dim,
+    num_patch=(num_patch_x, num_patch_y),
+    num_heads=num_heads,
+    window_size=window_size,
+    shift_size=shift_size,
+    num_mlp=num_mlp,
+    qkv_bias=qkv_bias,
+    drop_rate=drop_rate,
+)(x)
+x = PatchMerging((num_patch_x, num_patch_y), embed_dim=embed_dim)(x)
+x = GlobalAveragePooling1D()(x)
+output = layers.Dense(num_classes, activation="softmax")(x)
+
+"""
+### Train on CIFAR-10
+
+We will now finally train the model on CIFAR-10.
+"""
+
+model = keras.Model(input, output)
+model.compile(
+    loss=keras.losses.categorical_crossentropy,
+    optimizer=keras.optimizers.Adam(learning_rate=learning_rate, clipvalue=clipvalue),
+    metrics=[
+        keras.metrics.CategoricalAccuracy(name="accuracy"),
+        keras.metrics.TopKCategoricalAccuracy(5, name="top-5-accuracy"),
+    ],
+)
+
+history = model.fit(
+    x_train,
+    y_train,
+    batch_size=batch_size,
+    epochs=num_epochs,
+    validation_split=validation_split,
+)
+
+"""
+Let's now visualize the training progress of the model.
+"""
+
+plt.plot(history.history["loss"], label="train_loss")
+plt.plot(history.history["val_loss"], label="val_loss")
+plt.xlabel("Epochs")
+plt.ylabel("Loss")
+plt.title("Train and Validation Losses Over Epochs", fontsize=14)
+plt.legend()
+plt.grid()
+plt.show()
+
+"""
+Let's see final results of the training on CIFAR-10
+"""
+
+loss, accuracy, top_5_accuracy = model.evaluate(x_test, y_test)
+print(f"Test loss: {round(loss, 2)}")
+print(f"Test accuracy: {round(accuracy * 100, 2)}%")
```

Comment:
That makes sense, as said in the previous comment I now use CIFAR-100 but as you mention I have also added:
- Increased dropout by a minuscule
- 2 Data Augmentation layers, we could most certainly add more and improve performance but I wanted to keep the training time in Colab ~30 minutes which I believe would not be super long
- Increased the number of epochs, but made other changes to compensate for the training time to be almost the same

This gets us to 75% top-5 accuracy on CIFAR-100 with no signs of overfitting so the readers could probably just increase the number of epochs. Do you think this would suffice?

Reply:
It's fine to train the model for fewer epochs than needed in order to keep the runtime short. See instructions here: https://github.com/keras-team/keras-io/blob/master/contributor_guide.md#demonstrate-the-full-power-of-your-model-but-keep-the-run-time-short

What do you get when you train the model to convergence? A reasonable accuracy (not top 5) on CIFAR 100 would be higher than 70%, for the full model (which may be trained on as much resources as you want)